---
title: "The Nh3D2.0 Vignette"
author:
  - name: Heewon (Judy) Lee
  - affiliation: 
    - http://orcid.org/0000-0002-1134-6758
    - University of Toronto
email: heewon.lee@mail.utoronto.ca
date: "`r Sys.Date()`"
abstract: >
  Nh3D2.0 is a non-homologous 3D protein dataset containing 
  the best representatives for each homology group referenced
  by the CATH database.
output:
  BiocStyle::html_document:
    toc: true
    toc_float: true
    fig_caption: yes
vignette: >
  %\VignetteIndexEntry{The Nh3D2.0 Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

<style>
  .add {
    color: #22863a;
    background-color: #f0fff4;
  }
  .del {
    color: #b31d28;
    background-color: #ffeef0;
  }
</style>

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)

# knitr knits in a new session with an empty global workspace after setting its
# working directory to ./vignettes. To make your package functions available in
# the vignette, you have to load the library. The following two lines should
# accomplish this without manual intervention:
pkgName <- trimws(gsub("^Package:", "", readLines("../DESCRIPTION")[1]))
library(pkgName, character.only = TRUE)

```

&nbsp;

<div style="font-size:90%;border:solid 1px #337ab7;background-color:#FAFBFC;margin-right:0px;">
There are many links and references in this document. If you find anything here
ambiguous, inaccurate, outdated, incomplete, or broken, please
[file an issue](https://github.com/judyheewonlee/Nh3D2.0/issues)!
</div>

&nbsp;

# About this Vignette

This is the Vignette for the Nh3D2.0 R package developed by Heewon Lee for a BCB430 Project. This project
is also supervised by Professor Boris Steipe. 

In this vignette, there will be guidelines as to how to use Nh3D2.0 and the process of developing the package.
References are listed in the `Bibliogrpahy and Citations` section below.

&nbsp;

# What is Nh3D 2.0?

&nbsp;

Nh3D2.0 is a non-homologous 3D protein dataset containing the best 
representatives for each homology group referenced by the CATH database.
It accesses the [CATH](http://www.cathdb.info/) database and the data
files in order to determine the best representatives in each architecture, 
topology or homology group. It is based off the original [**Nh3D**](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1182382/) program
published in 2005.

&nbsp;

## Why Nh3D 2.0 should be made

&nbsp;

Non-homologous datasets generated by different algorithms and databases are 
mostly dependant on simply determining sequence similarity between sequences. 
Most developers believe that if sequence similarity is below a certain percentage,
the two protein domains are considered to be **non-homologous** but can lead 
to inaccuracies. Some sequences may be homologous and have low sequence similarity due to an old common ancestor and may not necessarily have high sequence similarity.
In order to tackle this, we access the [CATH database](http://www.cathdb.info) which
segregates protein domains into classifications; class, architecture, topology and
homology. 


The CATH database is a hierarchical protein domain classification database. The protein structures
are taken from the [Protein Data Bank (PDB)](https://www.rcsb.org/) and are separated into 
individual structural domains. If there is sufficient evidence that a set of domains
share a clear common ancestor, they are grouped together into a single homologous  superfamily.

  - **Homologous Superfamilies** denoted by **H** level in CATH share the same fold
  - **Topology Superfamilies** denoted by **T** level in CATH share the same fold,  but do not have 
  sufficient evidence to suggest a clear evolutionary relationship. 
  - **Architecture Superfamilies** denoted by **A** level in the CATH are topologies that usually 
  share a similar spatial arrangement of secondary structures.
  - **Class Superfamilies** denoted by **C** level in the CATH are the 40 architecture that are separated into 
  four main classes based on he content of secondary structure (mainly alpha, mainly beta, mixed alpha-beta, few
  secondary structures)

The latest version of the CATH-Gene3D protein structure classification 
database provides annotations for over 235 000 protein domain structures and includes 
25 million domain predictions. 

By Using the CATH database and several algorithms, **Nh3D 2.0** aims to provide the user with 
the best representative protein domains in each homology group. 

There are several steps that are executed in order to determine the best 
representative protein domains in each homologoy group in CATH:

1. Retrieving Resolution and Sequence length
2. Retrieving RFactors
3. Retrieving Stereochemical Validation Data
4. Weighting the top representatives
5. Performing a tandem repeat finder to determine tandem repeats within a domain

&nbsp;

## Retrieving Resolution and Sequence Length

&nbsp;

Sequence length and resolution are retrieved from the **PDB** through the RESTful web services
provided by the database. This allows the retrieval of XML files which can contain data from
specified search parameters provided by the user. 

**Sequence length** should be considered when determining the best representative protein domain since longer sequences provide more information in general about the protein and it's structure compared to shorter sequences. 

Moreover, **resolution** is the most important factor in determining the best representative protein domain in a homology group. This is due to the fact that resolution is a measure of the quality of the data collected on a protein structure. Thus, the sequence with a lower resolution value (in Angstroms) is considered a good potential candidate. The only problem that occurs when evaluating sequences with simply resolution is **bias**. Researchers will tend to manually fit their protein structures during x-ray crystallography in order to develop a better resolution value. This obviously creates bias since the natural structure of the protein has a resolution value worse than the value found through modifying the protein's natural structure. In order to remove this bias from our evaluations when determining the best representative domains, the program must also take into consideration **Rfactors** and **Stereochemical Validation Scores**.

&nbsp;

## Retrieving R Factors

&nbsp;

**RFactors** are modified resolution values and are also commonly used as a measure of data quality for protein domains. 
In Nh3D2.0, **Rfree**, **RWork**, and **RObserved** are Rfactors that are also considered when determining the quality of protein data. 

- **Rfree** is a measure of the fit of the model to a small subset of the experimental data, which was not used in model refinement
- **RWork** is a residual factor for reflections that satisfy the resolution and observation limits and that were used as the working reflections in the refinement. This measurement does not include reflections set aside for cross-validation in a "free" R-factor set 
- **RObserved** is a residual factor for reflections that satisfy the resolution and observation limits. This quantity includes reflections set aside for cross-validation in a "free" R-factor set. 

In Nh3D2.0, these values are utilized in combination with the **Resolution** and **Stereochemical Validation Scores** in order to weigh the value of the protein domain data. It should be noted that Resolution takes precedence over the Rfactors, and the Rfactor values take precedent over Stereochemical Validation Data. 

&nbsp;

## Retrieving Stereochemical Validation Data

&nbsp;

Stereochemical validation data is added to the weighting system in combination with Resolution and Rfactor values to negate the effects of bias as stated before. If there is bias present in the Resolution or Rfactor values in the protein data, the stereochemical properties must have been altered in order to retrieve a better resolution value. Thus, Nh3D 2.0 accesses the [World Wide Protein Data Bank (wwPDB)](https://www.wwpdb.org/) to retrieve Stereochemical Validation scores including **Clashscore**, **Ramachandran outlier score**, **Sidechain outlier score** and **RSRZ outlier score**.

  - **Clashscore:** This score represents the number of pairs of atoms in the model that are unusually close to each other. It is calculated by MolProbity and expressed as the number or such clashes per thousand atoms.
- **Ramachandran Outlier Score:** A residue is considered to be a Ramachandran plot outlier if the combination of its φ and ψ torsion angles is unusual. This score is calculated as the percentage of Ramachandran outliers with respect to the total number of residues in the protein sequence.
- **Sidechain Outlier Score:** The sidechain outlier score is calculated as the percentage of residues with an unusual sidechain conformation with respect to the total number of residues in the structure.
- **RSRZ Outlier Score:** The real-space R-value (RSR) is a measure of the quality of fit between a part of an atomic model (in this case, one residue) and the data in real space. The RSR Z-score (RSRZ) is a normalization of RSR specific to a residue type and a resolution bin and is only calculated for standard amino acids and nucleotides in protein, DNA and RNA. A residue is considered an RSRZ outlier if its RSRZ value is greater than 2. The RSRZ outlier score is calculated as the percentage RSRZ outliers with respect to the total number of residues for which RSRZ was computed.

&nbsp;

## Weighting the top representatives

&nbsp;

To weight the top representative scores, Nh3D2.0 uses the **Multi-Objective Optimization** approach to determine statistically which protein domain has the best overall quality of data [5]. The weighting includes the Resolution, Rfactors, and Stereochemical validation data. Resolution and Rfactor values take precedent over Stereochemical validation data, and the Resolution values take slight precedence over the Rfactor values. If bias is present in the quality of protein data, the weight of the stereochemical data will be taken into consideration in the ranking process. 

&nbsp;

## Performing an internal repeat finder algorithm

&nbsp;

The last step in the computational process in determining the protein with the highest quality of data is determining for each domain if there are internal repeats within the sequence. If there are internal repeats, this would also cause slight bias due to the fact that internal repeats introduce longer sequence lengths and higher resolution values (internal duplications of the same sequence increases the length of the sequence but maintains the same resolution values). Therefore, sequences containing internal repeats should be removed from the potential candidates for representing a homology group. 

Note that this portion of the program is yet to be completed in future updates of Nh3D2.0. However, the user may still retrieve the top `n` representative protein domains in a specified CATH superfamily based off Resolution, Rfactor and Stereochemical Validation Score.

&nbsp;

## The Overall Workflow of Nh3D2.0

Below is a diagram of the package workflow, and the processes taken in order to determine the best representatives.

&nbsp;

![](../inst/img/Nh3D2.0Workflow.jpg?sanitize=true "Nh3D2.0 Package Workflow")

&nbsp;

# Data Semantics

&nbsp;

The resulting dataframe generated from Nh3D2.0 takes the following format:

- **ID**: The PDB ID for the protein entry
- **Length**: The length of the protein sequence according to the PDB
- **Resolution**: The resolution score of the protein sequence according to the PDB
- **Sequence**: The sequence of each protein entry retrieved from the PDB
- **Class**: The Class superfamily the protein belongs to according to the CATH database
- **Architecture**: The Architecture superfamily the protein belongs to according to the CATH database
- **Topology**: The Topology superfamily the protein belongs to according to the CATH database
- **Homology**: The Homology superfamily the protein belongs to according to the CATH database
- **Rfree**: The Rfree value retrieved from the PDB
- **RWork**: The RWork value retrieved from the PDB
- **RObserved**: The RObserved value retrieved from the PDB
- **Clashscore**: The clashscore value retrieved from wwPDB
- **Ramchandran**: The clashscore value retrieved from wwPDB
- **RSRZ**: The RSRZ outlier score retrieved from wwPDB
- **Sidechain**: The sidechain outlier score value retrieved from wwPDB
- **Rank**: The Rank of each entry determined by the weighting function in the package. **Note that the lower the value, the better the quality of the data**.

&nbsp;

Each entry in the dataframe can be accessed by running:

```R
cathTable$ID
cathTable$Length
cathTable$Class

# Where [...] is the annotation the user would like to access
cathTable$[...]

```

&nbsp;

# How to use Nh3D 2.0

&nbsp;

### 1.Data Setup

&nbsp;

In order to first start using Nh3D, we must set up the data by 
generating the cathTable. Use the function generateCathTable() in order to 
create the table.

```R
cathTable <- generateCathTable()

```

The cathTable contains the protein domains in each CATH identifier

&nbsp;

### 2. Data Update

&nbsp;

In order to update the cathTable and the data directly from the CATH database, 
the user may call the **`updateCath()`** function to update and/or 
create a new cathTable.

```R
cathTable <- updateCath()
```

&nbsp;

### 3. Viewing CATH Identifiers and descriptions

&nbsp;

If the user would like to retrieve some basic description and 
identifier for particular protein domains in the CATH database, they may 
generate a data frame containing the protein name, identifier 
and description.

```R
cathKeys <- generateCathKeyDF(cathTable)

(cathKeys)
```

Below is an image of the relational data model for the CATH Key datatable.
Since there is a hierarchy where Class > Architecture > Topology > Homology,
a relational datamodel was made to display and access the data.

&nbsp;

![](../inst/img/Nh3D2.0KeyModel.jpg?sanitize=true "Nh3D2.0 CATH Key Data Model")

&nbsp;

The code below retrieves a particular topology group
description, name and CATH identifier. 
Assume we would like to see the topology group with the 
**1.10.220** identifier

```R
topologyID <- "1.10.220"
cathKeys$Topology[topologyID == cathKeys$Topology$TopologyID,]
```

This returns the data row containing the name, description and 
CATH identifier. The description states that this group 
is part of **Annexin V, domain 1**.

&nbsp;

### 4.Retrieving the Non-homologous Dataset

&nbsp;

Assume we want to now retrieve the non-homologous protein
domains in the topology group from the previous example.

```R
cID <- 1
aID <- 10
tID <- 220

selCathTable <- Nh3D(cathTable, class = cID, architecture = aID, tID = 220)

(selCathTable)
```

The resulting data frame will contain protein domains that fall under the 
topology group according to the cath database. There are several 
annotations included in each entry, most importantly the ranking of
each protein domain. 

```R
getBestDomains(selCathTable, n = 10)
```

We retrieve the top 10 representative domains within the selected group.
With the ten representatives, we get the **PDB ID's** of each domain, 
the **length of the sequence**, **Resolution**, **Rfactors**, **Stereochemical**
**Validation Scores**, and the **Rank**.

&nbsp;


# Bibliographies and Citations

&nbsp;

1. [Nh3D: A reference dataset of non-homologous protein structures](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1182382/)

This was the original Nh3D program that was created in 2005 for C.
This source was used as a basic outline as to how Nh3D2.0 should be developed.
Unfortunately, the original Nh3D program is unavailable for use.

2. [R Package Structure](http://r-pkgs.had.co.nz/)

This package follows the structure and guidelines to R packages building
by Hadley Wickham.

3. [Tandem Repeats in Proteins: Prediction Algorithms and Biological Role](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4585158/)

This source links many different tandem repeat finder algorithms that
may be useful for future updates in Nh3D2.0. 

4. [CATH: comprehensive structural and functional annotations for genome sequences](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4384018/)

This is the official publication for the CATH database which is used as a reference
database in the Nh3D2.0 package.

5. [A tutorial on multiobjective optimization: fundamentals and evolutionary methods](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6105305/)

Multi-objective optimization is a statistical approach utilized in the Nh3D2.0 package
in order to determine the weighting/ranking of the protein domains in each group.

&nbsp;

# Further reading
<a name="a1"></a>

- [CATH database info](http://www.cathdb.info/wiki)
- [PDB101 - Resolution](https://pdb101.rcsb.org/learn/guide-to-understanding-pdb-data/resolution)
- [PDB101 - Rfactors](https://pdb101.rcsb.org/learn/guide-to-understanding-pdb-data/r-value-and-r-free)
- The [RStudio R markdown reference guide](https://www.rstudio.com/wp-content/uploads/2015/03/rmarkdown-reference.pdf)

# Session Info

```{r, echo=TRUE}
sessionInfo()
```


<!-- End -->
